cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,141 @@ public long getProgramRunCount(ProgramId programId) {
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Compute the old run count if the run record is older than first start up time after 5.1
+   */
+  boolean computeOldRunCount(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_STARTED, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_SUSPENDED, maxRows);
+    upgradeDone &= computeOldCompletedRunCount(maxRows);
+    return upgradeDone;
+  }
+
+  /**
+   * Upgrades the rowkeys for the active run record. Note that we do not count the run record of the old active runs
+   * since it is hard to deal with the races and we expect user to stop all the programs before they do the upgrade.
+   *
+   * @param recordType type of the record
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if no rows required an upgrade
+   */
+  private boolean upgradeActiveRowkey(String recordType, int maxRows) {
+    LOG.info("Checking upgrade for {}", recordType);
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
+    Map<MDSKey, RunRecordMeta> oldMap = listKV(startKey, RunRecordMeta.class, maxRows);
+
+    for (Map.Entry<MDSKey, RunRecordMeta> oldEntry : oldMap.entrySet()) {
+      MDSKey oldKey = oldEntry.getKey();
+      RunRecordMeta runRecord = oldEntry.getValue();
+      // extract the program id from the old key, the format will be same after the program name is reached, so the
+      // method can be used directly.
+      MDSKey newKey = getProgramKeyBuilder(TYPE_RUN_RECORD_ACTIVE, getProgramID(oldKey))
+        .add(runRecord.getStartTs())
+        .add(runRecord.getPid())
+        .build();
+
+      delete(oldKey);
+      write(newKey, runRecord);
+    }
+
+    LOG.info("Upgrading {} active run records of {}", oldMap.size(), recordType);
+
+    // If we are not able to get the max number of rows, that means we reach the end of this record type
+    return oldMap.size() < maxRows;
+  }
+
+  /**
+   * Upgrades the rowkeys for the completed run record which is older than the time at row key
+   * RUN_COUNT_FIRST_UPGRADE_TIME
+   *
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if all old completed run records has been scanned
+   */
+  private boolean computeOldCompletedRunCount(int maxRows) {
+    LOG.info("Starting to compute the old run count");
+    Long upgradeStart = get(new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME), Long.TYPE);
+    if (upgradeStart == null) {
+      // this should not happen
+      LOG.warn("Unable to get the first upgrade start time, will use the current timestamp to distinguish the old " +
+                 "run records");
+    }
+
+    long upgradeStartTime =
+      upgradeStart == null ? TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()) : upgradeStart;
```

Comment:
probably easier just to store the start time in seconds.

Reply:
We already store it in seconds, this is to prevent a NPE where we are not able to get the upgrade start time from the store. It should not happen since we always write the start time before any step of the ugprade